roboteam_ai/src/coach/defence/DefencePositionCoach.cpp

```diff
@@ -322,6 +327,65 @@ Vector2 DefencePositionCoach::findPositionForBlockBall(const Line &blockLine) {
     }
     return getPosOnLine(blockLine, 0.3);
 }
+
+world::WorldData DefencePositionCoach::getTheirAttackers(const world::WorldData &world) {
+    std::vector<world::Robot> theirAttackers;
+    for (const world::Robot &robot :world.them) {
+        // we remove any attackers that are outside of the field or in our defence area
+        if (world::field->pointIsInDefenceArea(robot.pos, true, 0.04)) {
+            continue;
+        }
+        else if (! world::field->pointIsInField(robot.pos, 0.1)) {
+            continue;
+        }
+        else {
+            theirAttackers.push_back(robot);
+        }
+    }
+    world::WorldData newWorld = world;
+    newWorld.them = theirAttackers;
+    return newWorld;
+}
+bool DefencePositionCoach::validNewPosition(const Vector2 &position, const world::WorldData &world) {
+    double collisionRadius = 0.15;// a little smaller than 2 robot radii so that we can make solid walls still
+    for (const auto &robot: world.us) {
+        if ((robot.pos - position).length() < collisionRadius) {
+            return false;
+        }
+    }
+    return true;
+}
+std::shared_ptr<double> DefencePositionCoach::pickNewPosition(const Line& line,
+        const world::WorldData &world) {
+    double lineLen=(line.second-line.first).length();
+    double minSkipDist=0.05;
+    double skipDist=max(minSkipDist,lineLen/30.0);
+    // try positions on the line by skipping forward skipDist every time. If one works return
+    for (int aggFac = 0; aggFac <1/skipDist; ++ aggFac) {
```

Comment:
`aggFac`?

Reply:
this is a whole bunch of magic here

Reply:
read the comment, it should be quite clear; we pick a position on the line by trying out positions from the start and checking every time whether or not they are valid. aggFac=aggressionFactor (renamed this)

Reply:
I added a small comment